### PR TITLE
Do not merge - alternate fix for issue 8351 kr

### PR DESCRIPTION
**Pull request by KevinRansom**
Contains @smoothdeveloper 's fix and test updates. 

Includes :
-  updated member access validation
- Additional testing

### CODE REVIEW

File: src/fsharp/AccessibilityLogic.fs [hunk 1]
```diff
@@ -249,12 +249,35 @@ let private IsILMethInfoAccessible g amap m adType ad ilminfo =
...
+        match pdef.GetMethod, pdef.SetMethod with 
+        | Some mref, None 
+        | None, Some mref -> (resolveILMethodRef tdef mref).Access
+
+        | Some mrefGet, Some mrefSet ->
+            let getA = (resolveILMethodRef tdef mrefGet).Access
+            let setA = (resolveILMethodRef tdef mrefSet).Access
+
+            // use the accessors to determine the visibility of the property
+            match getA, setA with
```
Comment (TIHan) [inline src/fsharp/AccessibilityLogic.fs:261]: This looks really good. The order to determine most visible is now determined here and should give us the same effect.

Comment (TIHan) [reply src/fsharp/AccessibilityLogic.fs:261]: I'm in favor of this rather than having a specific order for the DU.

Comment (TIHan) [reply src/fsharp/AccessibilityLogic.fs:261]: My only suggestion is the follow the ordering that was dictated by Roslyn:
```
   | CompilerControlled
    | Private
    | FamilyAndAssembly // protected and internal
    | Family // protected
    | Assembly // internal
    | FamilyOrAssembly // protected or internal
    | Public
```

Comment (KevinRansom) [reply src/fsharp/AccessibilityLogic.fs:261]: I assume you are suggesting adding compiler controlled ... I can for sure do it, although for the life of me I have no idea what it even means for property accessors in F# :-)

Comment (smoothdeveloper) [reply src/fsharp/AccessibilityLogic.fs:261]: I'm still finding this solution has more moving pieces than the original with the exhaustive test I added that Jon asked.

In anycase, it is good to have more top level testing that are added here.

I think you know better what fits the best in the compiler maintenance mindset, main thing is to have the bug fixed in a future release :)

Maybe a comment on the match that order of cases is critical to remain "from most visible to least visible" wouldn't hurt.

Comment (KevinRansom) [reply src/fsharp/AccessibilityLogic.fs:261]: @smoothdeveloper , good call on the comment, I will add one.  The test case for the DU was certainly important if we were to rely on DU order it for this feature, however, we are not depending on that specific ordering with this change.

Comment (KevinRansom) [reply src/fsharp/AccessibilityLogic.fs:274]: added

File: src/fsharp/AccessibilityLogic.fs [hunk 2]
```diff
@@ -274,7 +287,10 @@ let GetILAccessOfILPropInfo (ILPropInfo(tinfo, pdef)) =
             | ILMemberAccess.FamilyAndAssembly, _
             | _, ILMemberAccess.FamilyAndAssembly -> ILMemberAccess.FamilyAndAssembly
 
-            | _ -> ILMemberAccess.Private
+            | ILMemberAccess.Private, _
+            | _, ILMemberAccess.Private -> ILMemberAccess.Private
+
+            | _ -> ILMemberAccess.CompilerControlled
```
Comment (smoothdeveloper) [inline src/fsharp/AccessibilityLogic.fs:293]:
```suggestion
            | ILMemberAccess.CompilerControlled, _
            | _, ILMemberAccess.CompilerControlled -> ILMemberAccess.CompilerControlled
```

this way, an addition will trigger a warning on this match expression lacking new cases.

File: src/fsharp/AccessibilityLogic.fs [hunk 3]
```diff
@@ -249,12 +249,51 @@ let private IsILMethInfoAccessible g amap m adType ad ilminfo =
...
+            //  ILMemberAccess.Assembly
+            //  ILMemberAccess.Family
+            //  ILMemberAccess.FamilyAndAssembly
+            //  ILMemberAccess.Private
+            //  ILMemberAccess.CompilerControlled
+            //          
+            let getA = (resolveILMethodRef tdef mrefGet).Access
+            let setA = (resolveILMethodRef tdef mrefSet).Access
+
+            // use the accessors to determine the visibility of the property
```
Comment (smoothdeveloper) [inline src/fsharp/AccessibilityLogic.fs:273]:
```suggestion
            // Use the accessors to determine the visibility of the property.
            // N.B. It is critical to keep the ordering in decreasing visibility order in the following match expression
```

**Comment by KevinRansom** [comment]
Pushed updates back to: source pr : https://github.com/dotnet/fsharp/pull/2867/files

**Comment by smoothdeveloper** [comment]
@KevinRansom somehow I don't see https://github.com/dotnet/fsharp/pull/9445/commits/ee6fd8ed754abb0883ee8a59f2cc4b280a02f88e in the source PR? is it possible to integrate it as well?

**Comment by KevinRansom** [comment]
Let me take a look, it should be there.

**Comment by KevinRansom** [comment]
@smoothdeveloper  ... bugger it got lost.

**Comment by KevinRansom** [comment]
Updated: https://github.com/dotnet/fsharp/pull/8354/commits/24b4d2823af9a4b301fca1fd36467a7f216db80e

**Pull request closed without merging**